File: _lib/entry-surface.mjs
// PCEF P2 entry surface collector.
//
// This artifact separates "known entry files" from later reachability math.
// It is evidence for module BFS and ranking confidence, not a direct fix
// recommendation.

import path from 'node:path';

import { producerMetaBase } from './artifacts.mjs';
import { buildAliasMap } from './alias-map.mjs';
import { collectFiles } from './collect-files.mjs';
import {
  collectHtmlModuleEntrypoints,
  collectPackagePublicSurfaceFiles,
  collectScriptEntrypoints,
  indexPublicSurfaceEntries,
} from './public-surface.mjs';
import { makeResolver } from './resolver-core.mjs';
import { isResolvedFile } from './resolver-core.mjs';
import { buildSubmoduleResolver } from './paths.mjs';
import {
  ACTION_MUTE,
  classifyFrameworkPolicy,
  createFrameworkPolicyContextForRepo,
  isConfigFile,
} from './classify-policies.mjs';
import { JS_FAMILY_LANGS } from './lang.mjs';

export const ENTRY_SURFACE_REQUEST_SCHEMA_VERSION = 'lumin-entry-surface-producer-request.v1';

function normalizeRel(relPath) {
  return String(relPath ?? '').replace(/\\/g, '/');
}

function sortedSet(set) {
  return [...set].sort((a, b) => a.localeCompare(b));
}

function sortedEvidenceObject(evidenceByFile) {
  const out = {};
  for (const file of [...evidenceByFile.keys()].sort((a, b) => a.localeCompare(b))) {
    out[file] = [...evidenceByFile.get(file)];
  }
  return out;
}

function sortedRecords(records) {
  return [...(records ?? [])].sort((a, b) =>
    String(a.packageDir ?? a.htmlFile ?? '').localeCompare(String(b.packageDir ?? b.htmlFile ?? '')) ||
    String(a.scriptName ?? '').localeCompare(String(b.scriptName ?? '')) ||
    String(a.reason ?? '').localeCompare(String(b.reason ?? '')) ||
    String(a.src ?? '').localeCompare(String(b.src ?? '')) ||
    String(a.resolvedFile ?? '').localeCompare(String(b.resolvedFile ?? '')));
}

function addEvidenceFile(targetSet, evidenceMap, relPath, evidence) {
  const rel = normalizeRel(relPath);
  targetSet.add(rel);
  if (evidence !== undefined) {
    if (!evidenceMap.has(rel)) evidenceMap.set(rel, []);
    evidenceMap.get(rel).push(evidence);
  }
}

function addFileVariants(targetSet, evidenceMap, relPath, evidence) {
  const rel = normalizeRel(relPath);
  const variants = new Set([rel]);
  if (/\.tsx$/.test(rel)) variants.add(rel.replace(/\.tsx$/, '.jsx'));
  else if (/\.jsx$/.test(rel)) variants.add(rel.replace(/\.jsx$/, '.tsx'));
  else if (/\.ts$/.test(rel) && !/\.d\.[cm]?ts$/.test(rel)) {
    variants.add(rel.replace(/\.ts$/, '.js'));
  } else if (/\.js$/.test(rel)) {
    variants.add(rel.replace(/\.js$/, '.ts'));
  }
  for (const variant of variants) addEvidenceFile(targetSet, evidenceMap, variant, evidence);
}

function collectKnownFiles({ root, symbolsData, includeTests, exclude }) {
  const files = new Set();
  try {
    for (const file of collectFiles(root, {
      includeTests,
      exclude,
      languages: JS_FAMILY_LANGS,
    })) {
      files.add(path.relative(root, file).replace(/\\/g, '/'));
    }
  } catch {
    // Entry surface is advisory evidence. If walking fails, keep going with
    // files already visible in symbols.json.
  }

  for (const file of Object.keys(symbolsData?.defIndex ?? {})) files.add(normalizeRel(file));
  for (const file of Object.keys(symbolsData?.reExportsByFile ?? {})) files.add(normalizeRel(file));
  for (const edge of symbolsData?.resolvedInternalEdges ?? []) {
    if (edge?.from) files.add(normalizeRel(edge.from));
    if (edge?.to) files.add(normalizeRel(edge.to));
  }

  return [...files].sort((a, b) => a.localeCompare(b));
}

function addIndexedEntries(indexedEntries, targetSet, evidenceByFile) {
  for (const [rel, evidence] of indexedEntries) {
    for (const item of evidence) addFileVariants(targetSet, evidenceByFile, rel, item);
  }
}

function addConcreteIndexedEntries(indexedEntries, targetSet, evidenceByFile) {
  for (const [rel, evidence] of indexedEntries) {
    for (const item of evidence) addEvidenceFile(targetSet, evidenceByFile, rel, item);
  }
}

function collectPublicApi({ root, repoMode, symbolsData, aliasMap, resolve }) {
  const files = new Set();
  const evidenceByFile = new Map();
  const addPublicApiVariants = (relPath, evidence) =>
    addFileVariants(files, evidenceByFile, relPath, evidence);

  addIndexedEntries(
    indexPublicSurfaceEntries(collectPackagePublicSurfaceFiles({ root, repoMode })),
    files,
    evidenceByFile,
  );

  for (const [spec, entry] of aliasMap) {
    if (typeof spec === 'string' && spec.startsWith('#')) continue;
    if (entry.source === 'imports') continue;
    if (entry.type === 'exact' && entry.path) {
      const rel = path.relative(root, entry.path).replace(/\\/g, '/');
      addPublicApiVariants(rel, {
        source: 'alias-map.exact',
        aliasSource: entry.source ?? null,
        resolvedFile: rel,
      });
    }
  }

  let transitiveAdded = 0;
  const reExportsByFile = symbolsData?.reExportsByFile ?? {};
  if (files.size > 0 && Object.keys(reExportsByFile).length > 0) {
    const visited = new Set();
    const queue = [...files].filter((p) => reExportsByFile[p] !== undefined || files.has(p));
    while (queue.length) {
      const current = queue.shift();
      if (visited.has(current)) continue;
      visited.add(current);
      const reExports = reExportsByFile[current];
      if (!reExports) continue;
      for (const r of reExports) {
        if (!r.source) continue;
        const resolved = resolve(path.join(root, current), r.source);
        if (!isResolvedFile(resolved)) continue;
        const rel = path.relative(root, resolved).replace(/\\/g, '/');
        if (!files.has(rel)) {
          addPublicApiVariants(rel, {
            source: 'public-reexport',
            fromFile: current,
            specifier: r.source,
            resolvedFile: rel,
          });
          transitiveAdded++;
        }
        if (!visited.has(rel)) queue.push(rel);
      }
    }
  }

  return { files, evidenceByFile, transitiveAdded };
}

function collectScriptEntries({ root, repoMode }) {
  const files = new Set();
  const evidenceByFile = new Map();
  const script = collectScriptEntrypoints({ root, repoMode });
  addIndexedEntries(
    indexPublicSurfaceEntries(script.entries),
    files,
    evidenceByFile,
  );
  return {
    files,
    evidenceByFile,
    unsupported: script.unsupported,
    unsupportedRawCount: script.unsupportedRawCount,
    unsupportedSampleLimit: script.unsupportedSampleLimit,
  };
}

function collectHtmlEntries({ root, repoMode, includeTests, exclude }) {
  const files = new Set();
  const evidenceByFile = new Map();
  const html = collectHtmlModuleEntrypoints({ root, repoMode, includeTests, exclude });
  addConcreteIndexedEntries(
    indexPublicSurfaceEntries(html.entries),
    files,
    evidenceByFile,
  );
  return { files, evidenceByFile, unresolved: html.unresolved };
}

function collectConfigEntries({ knownFiles }) {
  const files = new Set();
  const evidenceByFile = new Map();
  for (const file of knownFiles) {
    if (!isConfigFile(file)) continue;
    addEvidenceFile(files, evidenceByFile, file, { source: 'config-file-convention' });
  }
  return { files, evidenceByFile };
}

function collectFrameworkEntries({ root, repoMode, symbolsData, includeTests, exclude, knownFiles }) {
  const files = new Set();
  const evidenceByFile = new Map();
  const context = createFrameworkPolicyContextForRepo({
    root,
    repoMode,
    symbolsData,
    deadList: [],
    includeTests,
    exclude,
  });

  for (const file of knownFiles) {
    const defs = symbolsData?.defIndex?.[file] ?? {};
    const exportNames = Object.keys(defs);
    const candidates = exportNames.length > 0 ? exportNames : ['default'];
    for (const exportName of candidates) {
      const decision = classifyFrameworkPolicy(context, { file, exportName });
      if (decision.action !== ACTION_MUTE) continue;
      addEvidenceFile(files, evidenceByFile, file, {
        source: 'framework-policy',
        framework: decision.framework,
        ruleId: decision.ruleId,
        reason: decision.reason,
        evidence: decision.evidence ?? null,
      });
      break;
    }
  }

  return { files, evidenceByFile };
}

function mergeEvidenceMaps(...maps) {
  const merged = new Map();
  for (const map of maps) {
    for (const [file, evidence] of map) {
      if (!merged.has(file)) merged.set(file, []);
      merged.get(file).push(...evidence);
    }
  }
  return merged;
}

function parseErrorCount(symbolsData) {
  return (symbolsData?.meta?.warnings ?? [])
    .filter((w) => w?.code === 'parse-errors' || w?.kind === 'parse-errors' || w?.type === 'parse-errors')
    .reduce((sum, w) => sum + (Number(w?.count) || 0), 0);
}

function completenessLabels({ entryFiles, knownFiles, symbolsData, submoduleOf, limitations = [] }) {
  const parseErrors = parseErrorCount(symbolsData);
  const globalCompleteness = parseErrors > 0 || limitations.length > 0 ? 'medium' : 'high';
  const submodules = new Set([
    ...knownFiles.map(submoduleOf),
    ...entryFiles.map(submoduleOf),
  ]);
  const completenessBySubmodule = {};
  for (const submodule of [...submodules].sort((a, b) => a.localeCompare(b))) {
    completenessBySubmodule[submodule] = globalCompleteness;
  }
  return { globalCompleteness, completenessBySubmodule };
}

function collectEntrySurfaceParts({
  root,
  repoMode,
  symbolsData,
  includeTests = true,
  exclude = [],
}) {
  const aliasMap = buildAliasMap(root, repoMode, { exclude });
  const resolve = makeResolver(root, aliasMap);
  const submoduleOf = buildSubmoduleResolver(root, repoMode);
  const knownFiles = collectKnownFiles({ root, symbolsData, includeTests, exclude });

  const publicApi = collectPublicApi({ root, repoMode, symbolsData, aliasMap, resolve });
  const script = collectScriptEntries({ root, repoMode });
  const html = collectHtmlEntries({ root, repoMode, includeTests, exclude });
  const framework = collectFrameworkEntries({ root, repoMode, symbolsData, includeTests, exclude, knownFiles });
  const config = collectConfigEntries({ knownFiles });

  const entryFiles = sortedSet(new Set([
    ...publicApi.files,
    ...script.files,
    ...html.files,
    ...framework.files,
    ...config.files,
  ]));
  const evidenceByFile = mergeEvidenceMaps(
    publicApi.evidenceByFile,
    script.evidenceByFile,
    html.evidenceByFile,
    framework.evidenceByFile,
    config.evidenceByFile,
  );
  const unresolvedHtmlEntrypoints = sortedRecords(html.unresolved);
  const unsupportedScriptEntrypoints = sortedRecords(script.unsupported);

  return {
    knownFiles,
    publicApi,
    script,
    html,
    framework,
    config,
    entryFiles,
    evidenceByFile,
    unresolvedHtmlEntrypoints,
    unsupportedScriptEntrypoints,
    submoduleOf,
  };
}

function laneFacts(lane) {
  return {
    files: sortedSet(lane.files),
    evidenceByFile: sortedEvidenceObject(lane.evidenceByFile),
  };
}

function submoduleFacts(files, submoduleOf) {
  const out = {};
  for (const file of sortedSet(new Set(files))) {
    out[file] = submoduleOf(file);
  }
  return out;
}

export function collectEntrySurfaceFacts({
  root,
  repoMode,
  symbolsData,
  includeTests = true,
  exclude = [],
}) {
  const parts = collectEntrySurfaceParts({ root, repoMode, symbolsData, includeTests, exclude });
  return {
    schemaVersion: ENTRY_SURFACE_REQUEST_SCHEMA_VERSION,
    root,
    generated: new Date().toISOString(),
    includeTests,
    knownFiles: parts.knownFiles,
    parseErrorCount: parseErrorCount(symbolsData),
    submoduleByFile: submoduleFacts(
      [...parts.knownFiles, ...parts.entryFiles],
      parts.submoduleOf,
    ),
    publicApi: {
      ...laneFacts(parts.publicApi),
      transitiveAdded: parts.publicApi.transitiveAdded,
    },
    script: {
      ...laneFacts(parts.script),
      unsupportedRawCount: parts.script.unsupportedRawCount,
      unsupportedSampleLimit: parts.script.unsupportedSampleLimit,
      unsupported: parts.unsupportedScriptEntrypoints,
    },
    html: {
      ...laneFacts(parts.html),
      unresolved: parts.unresolvedHtmlEntrypoints,
    },
    framework: laneFacts(parts.framework),
    config: laneFacts(parts.config),
  };
}

export function buildEntrySurfaceArtifact({
  root,
  repoMode,
  symbolsData,
  includeTests = true,
  exclude = [],
}) {
  const parts = collectEntrySurfaceParts({ root, repoMode, symbolsData, includeTests, exclude });
  const { globalCompleteness, completenessBySubmodule } =
    completenessLabels({
      entryFiles: parts.entryFiles,
      knownFiles: parts.knownFiles,
      symbolsData,
      submoduleOf: parts.submoduleOf,
      limitations: parts.unresolvedHtmlEntrypoints,
    });

  return {
    meta: {
      ...producerMetaBase({ tool: 'build-entry-surface.mjs', root }),
      schemaVersion: 'entry-surface.v1',
      supports: {
        publicApiFiles: true,
        scriptEntrypointFiles: true,
        unsupportedScriptEntrypoints: true,
        htmlEntrypointFiles: true,
        unresolvedHtmlEntrypoints: true,
        frameworkEntrypointFiles: true,
        configEntrypointFiles: true,
        submoduleCompleteness: true,
      },
      includeTests,
      transitivePublicReexports: parts.publicApi.transitiveAdded,
      knownFileCount: parts.knownFiles.length,
    },
    publicApiFiles: sortedSet(parts.publicApi.files),
    scriptEntrypointFiles: sortedSet(parts.script.files),
    unsupportedScriptEntrypointCount: parts.script.unsupportedRawCount,
    unsupportedScriptEntrypointSampleLimit: parts.script.unsupportedSampleLimit,
    unsupportedScriptEntrypoints: parts.unsupportedScriptEntrypoints,
    htmlEntrypointFiles: sortedSet(parts.html.files),
    unresolvedHtmlEntrypoints: parts.unresolvedHtmlEntrypoints,
    frameworkEntrypointFiles: sortedSet(parts.framework.files),
    configEntrypointFiles: sortedSet(parts.config.files),
    entryFiles: parts.entryFiles,
    evidenceByFile: sortedEvidenceObject(parts.evidenceByFile),
    globalCompleteness,
    completenessBySubmodule,
  };
}

File: skills/lumin-repo-lens-lab/_engine/rust/rust-main/lumin-audit-core/src/entry_surface.rs
use anyhow::{bail, Result};
use serde::Deserialize;
use serde_json::{json, Value};
use std::collections::{BTreeMap, BTreeSet};

pub const ENTRY_SURFACE_SCHEMA_VERSION: &str = "entry-surface.v1";
pub const ENTRY_SURFACE_REQUEST_SCHEMA_VERSION: &str = "lumin-entry-surface-producer-request.v1";

const TOOL_NAME: &str = "build-entry-surface.mjs";

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EntrySurfaceRequest {
    pub schema_version: String,
    pub root: String,
    pub generated: String,
    #[serde(default)]
    pub include_tests: bool,
    #[serde(default)]
    pub known_files: Vec<String>,
    #[serde(default)]
    pub parse_error_count: usize,
    #[serde(default)]
    pub submodule_by_file: BTreeMap<String, String>,
    #[serde(default)]
    pub public_api: PublicApiFacts,
    #[serde(default)]
    pub script: ScriptFacts,
    #[serde(default)]
    pub html: HtmlFacts,
    #[serde(default)]
    pub framework: EntryLaneFacts,
    #[serde(default)]
    pub config: EntryLaneFacts,
}

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EntryLaneFacts {
    #[serde(default)]
    pub files: Vec<String>,
    #[serde(default)]
    pub evidence_by_file: BTreeMap<String, Vec<Value>>,
}

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PublicApiFacts {
    #[serde(default)]
    pub files: Vec<String>,
    #[serde(default)]
    pub evidence_by_file: BTreeMap<String, Vec<Value>>,
    #[serde(default)]
    pub transitive_added: usize,
}

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScriptFacts {
    #[serde(default)]
    pub files: Vec<String>,
    #[serde(default)]
    pub evidence_by_file: BTreeMap<String, Vec<Value>>,
    #[serde(default)]
    pub unsupported_raw_count: usize,
    #[serde(default)]
    pub unsupported_sample_limit: usize,
    #[serde(default)]
    pub unsupported: Vec<Value>,
}

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HtmlFacts {
    #[serde(default)]
    pub files: Vec<String>,
    #[serde(default)]
    pub evidence_by_file: BTreeMap<String, Vec<Value>>,
    #[serde(default)]
    pub unresolved: Vec<Value>,
}

pub fn build_entry_surface_artifact(request: EntrySurfaceRequest) -> Result<Value> {
    if request.schema_version != ENTRY_SURFACE_REQUEST_SCHEMA_VERSION {
        bail!(
            "entry-surface-artifact: unsupported schemaVersion '{}'",
            request.schema_version
        );
    }

    let known_files = sorted_files(request.known_files);
    let public_api_files = sorted_files(request.public_api.files.clone());
    let script_files = sorted_files(request.script.files.clone());
    let html_files = sorted_files(request.html.files.clone());
    let framework_files = sorted_files(request.framework.files.clone());
    let config_files = sorted_files(request.config.files.clone());

    let mut entry_file_set = BTreeSet::new();
    extend_set(&mut entry_file_set, &public_api_files);
    extend_set(&mut entry_file_set, &script_files);
    extend_set(&mut entry_file_set, &html_files);
    extend_set(&mut entry_file_set, &framework_files);
    extend_set(&mut entry_file_set, &config_files);
    let entry_files = entry_file_set.into_iter().collect::<Vec<_>>();

    let mut evidence_by_file = BTreeMap::new();
    merge_evidence(&mut evidence_by_file, request.public_api.evidence_by_file);
    merge_evidence(&mut evidence_by_file, request.script.evidence_by_file);
    merge_evidence(&mut evidence_by_file, request.html.evidence_by_file);
    merge_evidence(&mut evidence_by_file, request.framework.evidence_by_file);
    merge_evidence(&mut evidence_by_file, request.config.evidence_by_file);

    let unsupported_script_entrypoints = sorted_records(request.script.unsupported);
    let unresolved_html_entrypoints = sorted_records(request.html.unresolved);
    let global_completeness =
        if request.parse_error_count > 0 || !unresolved_html_entrypoints.is_empty() {
            "medium"
        } else {
            "high"
        };
    let completeness_by_submodule = completeness_by_submodule(
        &known_files,
        &entry_files,
        &request.submodule_by_file,
        global_completeness,
    );

    Ok(json!({
        "meta": {
            "tool": TOOL_NAME,
            "generated": request.generated,
            "root": request.root,
            "schemaVersion": ENTRY_SURFACE_SCHEMA_VERSION,
            "supports": {
                "publicApiFiles": true,
                "scriptEntrypointFiles": true,
                "unsupportedScriptEntrypoints": true,
                "htmlEntrypointFiles": true,
                "unresolvedHtmlEntrypoints": true,
                "frameworkEntrypointFiles": true,
                "configEntrypointFiles": true,
                "submoduleCompleteness": true,
            },
            "includeTests": request.include_tests,
            "transitivePublicReexports": request.public_api.transitive_added,
            "knownFileCount": known_files.len(),
        },
        "publicApiFiles": public_api_files,
        "scriptEntrypointFiles": script_files,
        "unsupportedScriptEntrypointCount": request.script.unsupported_raw_count,
        "unsupportedScriptEntrypointSampleLimit": request.script.unsupported_sample_limit,
        "unsupportedScriptEntrypoints": unsupported_script_entrypoints,
        "htmlEntrypointFiles": html_files,
        "unresolvedHtmlEntrypoints": unresolved_html_entrypoints,
        "frameworkEntrypointFiles": framework_files,
        "configEntrypointFiles": config_files,
        "entryFiles": entry_files,
        "evidenceByFile": evidence_by_file,
        "globalCompleteness": global_completeness,
        "completenessBySubmodule": completeness_by_submodule,
    }))
}

fn normalize_rel(path: impl AsRef<str>) -> String {
    path.as_ref().replace('\\', "/")
}

fn sorted_files(files: Vec<String>) -> Vec<String> {
    files
        .into_iter()
        .map(normalize_rel)
        .collect::<BTreeSet<_>>()
        .into_iter()
        .collect()
}

fn extend_set(target: &mut BTreeSet<String>, files: &[String]) {
    for file in files {
        target.insert(file.clone());
    }
}

fn merge_evidence(
    target: &mut BTreeMap<String, Vec<Value>>,
    evidence: BTreeMap<String, Vec<Value>>,
) {
    for (file, values) in evidence {
        target
            .entry(normalize_rel(file))
            .or_default()
            .extend(values);
    }
}

fn sorted_records(records: Vec<Value>) -> Vec<Value> {
    let mut records = records;
    records.sort_by_key(record_key);
    records
}

fn record_key(record: &Value) -> [String; 5] {
    [
        record_string(record, "packageDir")
            .or_else(|| record_string(record, "htmlFile"))
            .unwrap_or_default(),
        record_string(record, "scriptName").unwrap_or_default(),
        record_string(record, "reason").unwrap_or_default(),
        record_string(record, "src").unwrap_or_default(),
        record_string(record, "resolvedFile").unwrap_or_default(),
    ]
}

fn record_string(record: &Value, field: &str) -> Option<String> {
    record.get(field).map(|value| match value {
        Value::String(text) => text.clone(),
        Value::Null => String::new(),
        other => other.to_string(),
    })
}

fn completeness_by_submodule(
    known_files: &[String],
    entry_files: &[String],
    submodule_by_file: &BTreeMap<String, String>,
    completeness: &str,
) -> BTreeMap<String, String> {
    let mut submodules = BTreeSet::new();
    for file in known_files.iter().chain(entry_files.iter()) {
        submodules.insert(
            submodule_by_file
                .get(file)
                .cloned()
                .unwrap_or_else(|| ".".to_string()),
        );
    }
    submodules
        .into_iter()
        .map(|submodule| (submodule, completeness.to_string()))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn request() -> Result<EntrySurfaceRequest> {
        Ok(serde_json::from_value(json!({
            "schemaVersion": ENTRY_SURFACE_REQUEST_SCHEMA_VERSION,
            "root": "C:/repo",
            "generated": "2026-07-04T00:00:00.000Z",
            "includeTests": false,
            "knownFiles": ["src\\api.ts", "src/page.tsx", "vite.config.ts"],
            "parseErrorCount": 0,
            "submoduleByFile": {
                "src/api.ts": "src",
                "src/cli.ts": "src",
                "src/page.tsx": "src",
                "vite.config.ts": "."
            },
            "publicApi": {
                "files": ["src/api.ts"],
                "transitiveAdded": 1,
                "evidenceByFile": {
                    "src/api.ts": [{ "source": "package-json.exports" }]
                }
            },
            "script": {
                "files": ["src/cli.ts"],
                "unsupportedRawCount": 2,
                "unsupportedSampleLimit": 20,
                "unsupported": [
                    { "packageDir": "pkg-b", "scriptName": "start", "reason": "shell" },
                    { "packageDir": "pkg-a", "scriptName": "dev", "reason": "unsupported" }
                ],
                "evidenceByFile": {
                    "src/cli.ts": [{ "source": "package-script" }]
                }
            },
            "html": {
                "files": ["src/page.tsx"],
                "unresolved": [{ "htmlFile": "index.html", "src": "./missing.ts" }],
                "evidenceByFile": {
                    "src/page.tsx": [{ "source": "html-module" }]
                }
            },
            "framework": {
                "files": ["src/page.tsx"],
                "evidenceByFile": {
                    "src/page.tsx": [{ "source": "framework-policy" }]
                }
            },
            "config": {
                "files": ["vite.config.ts"],
                "evidenceByFile": {
                    "vite.config.ts": [{ "source": "config-file-convention" }]
                }
            }
        }))?)
    }

    #[test]
    fn builds_entry_surface_projection_from_js_facts() -> Result<()> {
        let artifact = build_entry_surface_artifact(request()?)?;

        assert_eq!(artifact["meta"]["tool"], TOOL_NAME);
        assert_eq!(
            artifact["meta"]["schemaVersion"],
            ENTRY_SURFACE_SCHEMA_VERSION
        );
        assert_eq!(artifact["meta"]["includeTests"], false);
        assert_eq!(artifact["meta"]["knownFileCount"], 3);
        assert_eq!(artifact["meta"]["transitivePublicReexports"], 1);
        assert_eq!(
            artifact["entryFiles"],
            json!(["src/api.ts", "src/cli.ts", "src/page.tsx", "vite.config.ts"])
        );
        assert_eq!(artifact["globalCompleteness"], "medium");
        assert_eq!(artifact["completenessBySubmodule"]["."], "medium");
        assert_eq!(artifact["completenessBySubmodule"]["src"], "medium");
        assert_eq!(
            artifact["unsupportedScriptEntrypoints"][0]["packageDir"],
            "pkg-a"
        );
        assert_eq!(
            artifact["evidenceByFile"]["src/page.tsx"]
                .as_array()
                .map_or(0, Vec::len),
            2
        );
        Ok(())
    }

    #[test]
    fn high_completeness_when_no_parse_errors_or_unresolved_html() -> Result<()> {
        let mut request = request()?;
        request.html.unresolved.clear();
        let artifact = build_entry_surface_artifact(request)?;

        assert_eq!(artifact["globalCompleteness"], "high");
        Ok(())
    }

    #[test]
    fn rejects_unknown_request_schema() -> Result<()> {
        let mut request = request()?;
        request.schema_version = "other".to_string();
        let err = match build_entry_surface_artifact(request) {
            Ok(_) => panic!("schema should reject"),
            Err(error) => error.to_string(),
        };

        assert!(err.contains("unsupported schemaVersion"));
        Ok(())
    }
}
